### DISCUSSION TITLE
Reflection service - using dynamic file descriptor sets

### DISCUSSION THREAD

**Discussion by dado0583** [hyperium/tonic#1332] [Q&A]
I've added tonic_reflection 0.6 to my Cargo.toml and added the reflection service, but it looks like I need to load the file descriptor set from the tonic_build output. I don't have that to hand.

Some context:
* I have a multi language implementation of models and client code in one repo ([here](https://github.com/FinTekkers/ledger-models/))
* The auto-generated code gets deployed as a crate called ledger-models [here](https://crates.io/crates/ledger-models). 

I am now writing code in a private repo (broker-service), and need to add the reflection service:

```
    let service = tonic_reflection::server::Builder::configure()
        .register_encoded_file_descriptor_set(proto::FILE_DESCRIPTOR_SET)
        .build()
        .unwrap();
```

I don't currently build the file descriptor set via include_file_descriptor_set. Do I "have" to do that, and ship with my ledger-models crate?

I'd prefer to do it dynamically if possible from the Rust-generated files. E.g. the service rust [file](https://github.com/FinTekkers/ledger-models/blob/main/ledger-models-rust/fintekkers.services.position_service.rs) or request rust [file](https://github.com/FinTekkers/ledger-models/blob/main/ledger-models-rust/fintekkers.services.position_service.rs).

Any pointers as to whether I must ship the file descriptor set from the tonic_build, or is there a way to do this dynamically? 

Thanks in advance for any help!

**Comment by LucioFranco** [comment]
I think you want to use https://docs.rs/tonic-build/latest/tonic_build/struct.Builder.html#method.file_descriptor_set_path in the repo that contains the protobuf files and embed the fds into that code. I think you can use `include_bytes` cuz iirc its not actual rust code. This will then ship it with that code and it should be either writable to disk or passable to the reflection code.

**Reply by dado0583** [reply]
The challenge I had with that is the ledger models crate defines service definitions for about 8 services.

Multiple codebases import that crate so they know what to implement and how to communicate to other services. But they may expose only one service definition.

Using tonic_reflection it by default exposes all the service definitions so for now I copied the code and added a filter functionality.

Or put in pseudo code:

- service A project imports ledger-models
- service A loads the file descriptor from there
- it pushes it into the reflection service (modified version)
- reflection service exposes only a subset of services via reflection (modified version)

I'm going to see if I can filter the file descriptor set before loading the bytes into the reflection service but this modified version works fine.
